libvmaf/test/test_picture.c

```diff
@@ -65,9 +69,219 @@ static char *test_picture_data_alignment()
+
+    VmafPicture pic_6;
+    err = vmaf_picture_pool_request_picture(pic_pool, &pic_6);
+    mu_assert("problem during vmaf_picture_pool_request_picture", !err);
+    mu_assert("data[0] should have been allocated", pic_6.data[0]);
+    mu_assert("data[1] should not have been allocated", !pic_6.data[1]);
+    mu_assert("data[2] should not have been allocated", !pic_6.data[2]);
+    mu_assert("pix_fmt should be VMAF_PIX_FMT_YUV400P",
+              pic_6.pix_fmt == VMAF_PIX_FMT_YUV400P);
+
+    mu_assert("pic_6 should use the same data buffer as pic_2 did. "
+              "the underlying queue should be LIFO.",
+              pic_2_data_buf == pic_6.data[0]);
+
+    err |= vmaf_picture_unref(&pic_3);
+    err |= vmaf_picture_unref(&pic_4);
+    err |= vmaf_picture_unref(&pic_5);
+    err |= vmaf_picture_unref(&pic_6);
+    mu_assert("problem during vmaf_picture_unref", !err);
+
+    err = vmaf_picture_pool_close(pic_pool);
+    mu_assert("problem during vmaf_picture_pool_close", !err);
+
+    return NULL;
+}
+
+typedef struct MyThreadPoolData {
+    VmafPicturePool *pic_pool;
+    unsigned i;
+    useconds_t timeout;
+} MyThreadPoolData;
+
+
+static void request_picture_and_unref(void *data)
+{
+    MyThreadPoolData *my_thread_pool_data = data;
+    VmafPicturePool *pic_pool = my_thread_pool_data->pic_pool;
+    VmafPicture pic;
+    //fprintf(stderr, "request: %i\n", my_thread_pool_data->i);
+    vmaf_picture_pool_request_picture(pic_pool, &pic);
+    //fprintf(stderr, "usleep=%d: %i\n",
+    //        my_thread_pool_data->timeout, my_thread_pool_data->i);
+    usleep(my_thread_pool_data->timeout);
+    //fprintf(stderr, "unref: %i\n", my_thread_pool_data->i);
+    vmaf_picture_unref(&pic);
+}
+
+static char *test_picture_pool_threaded()
+{
+    int err;
+
+    MyCookie my_cookie = {
+        .data_sz = 1920*1080,
+        .pix_fmt = VMAF_PIX_FMT_YUV400P,
+    };
+
+    VmafPicturePoolConfig cfg = {
+        .pic_cnt = 4,
+        .cookie = &my_cookie,
+        .alloc_picture_callback = my_alloc_picture,
+        .free_picture_callback = my_free_picture,
+    };
+
+    VmafPicturePool *pic_pool;
+    err = vmaf_picture_pool_init(&pic_pool, cfg);
+    mu_assert("problem during vmaf_picture_pool_init", !err);
+
+    VmafThreadPool *thread_pool;
+    const unsigned n_threads = 4;
+    err = vmaf_thread_pool_create(&thread_pool, n_threads);
+    mu_assert("problem during vmaf_thread_pool_init", !err);
+
+    const unsigned n = n_threads * 8;
+    for (unsigned i = 0; i < n; i++) {
+        MyThreadPoolData my_thread_pool_data = {
+            .pic_pool = pic_pool,
+            .i = i,
+            .timeout = 100000 - (100000 * ((float)i / n)),
+        };
+        err = vmaf_thread_pool_enqueue(thread_pool, request_picture_and_unref,
+                                       &my_thread_pool_data,
+                                       sizeof(my_thread_pool_data));
+        mu_assert("problem during vmaf_thread_pool_enqueue", !err);
+    }
+
```

Comment:
Pasting this snippet here makes my point above a bit clearer. Is it the desired functionality that this cookie changes ? 
```
    //ensure cookie is still MyCookie
    VmafPicture pic;
    vmaf_picture_pool_request_picture(pic_pool, &pic);
    VmafPicturePrivate* priv = pic.priv;
    err = memcmp(&my_cookie, priv->cookie, sizeof(my_cookie));
    mu_assert("my cookie is gone!", !err);

    vmaf_picture_unref(&pic);
```